File: backend/src/services/auth.service.ts
import bcrypt from 'bcrypt';
import jwt from 'jsonwebtoken';
import prisma from '../config/db';
import { BadRequestError, UnauthorizedError, NotFoundError } from '../utils/errors';
import { logAudit } from '../utils/audit';
import { Role, User } from '@prisma/client';

const JWT_SECRET = process.env.JWT_SECRET || 'dGhpcyBpcyBhIHZlcnkgbG9uZyBzZWNyZXQga2V5IGZvciBob3NwaXRhbCBtYW5hZ2VtZW50IHN5c3RlbSBqd3QgdG9rZW4gZ2VuZXJhdGlvbg==';
const ACCESS_EXPIRATION = parseInt(process.env.JWT_ACCESS_EXPIRATION || '900000', 10); // 15 mins
const REFRESH_EXPIRATION = parseInt(process.env.JWT_REFRESH_EXPIRATION || '604800000', 10); // 7 days

export interface AuthUserDto {
  id: number;
  username: string;
  email: string;
  firstName: string;
  lastName: string;
  phone: string | null;
  enabled: boolean;
  roles: string[];
  createdAt: Date;
  updatedAt: Date;
}

export const toUserDto = (user: User): AuthUserDto => ({
  id: user.id,
  username: user.username,
  email: user.email,
  firstName: user.firstName,
  lastName: user.lastName,
  phone: user.phone,
  enabled: user.enabled,
  roles: user.roles.map(r => `ROLE_${r}`),
  createdAt: user.createdAt,
  updatedAt: user.updatedAt,
});

/**
 * Generate Access Token
 */
export const generateAccessToken = (user: User): string => {
  return jwt.sign(
    {
      id: user.id,
      username: user.username,
      email: user.email,
      roles: user.roles,
    },
    JWT_SECRET,
    { expiresIn: `${ACCESS_EXPIRATION}ms` }
  );
};

/**
 * Generate and save Refresh Token
 */
export const createRefreshToken = async (userId: number): Promise<string> => {
  const token = jwt.sign({ id: userId }, JWT_SECRET, {
    expiresIn: `${REFRESH_EXPIRATION}ms`,
  });

  const expiryDate = new Date(Date.now() + REFRESH_EXPIRATION);

  await prisma.refreshToken.create({
    data: {
      token,
      userId,
      expiryDate,
    },
  });

  return token;
};

/**
 * Login Service
 */
export const loginUser = async (
  usernameOrEmail: string,
  password: string,
  ipAddress?: string
) => {
  const user = await prisma.user.findFirst({
    where: {
      OR: [{ username: usernameOrEmail }, { email: usernameOrEmail }],
    },
  });

  if (!user) {
    throw new UnauthorizedError('Invalid credentials');
  }

  // Account lock check
  if (user.accountLocked) {
    const isLockExpired =
      user.lockTime && new Date().getTime() - new Date(user.lockTime).getTime() > 30 * 60 * 1000;

    if (isLockExpired) {
      // Reset lock details
      await prisma.user.update({
        where: { id: user.id },
        data: {
          accountLocked: false,
          failedLoginAttempts: 0,
          lockTime: null,
        },
      });
    } else {
      throw new UnauthorizedError(
        'Account is locked due to multiple failed login attempts. Try again after 30 minutes.'
      );
    }
  }

  // Validate password
  const isMatch = await bcrypt.compare(password, user.passwordHash);
  if (!isMatch) {
    const nextAttempts = user.failedLoginAttempts + 1;
    const shouldLock = nextAttempts >= 5;

    await prisma.user.update({
      where: { id: user.id },
      data: {
        failedLoginAttempts: nextAttempts,
        accountLocked: shouldLock,
        lockTime: shouldLock ? new Date() : null,
      },
    });

    logAudit(user.id, 'LOGIN_FAILED', 'Failed login attempt', ipAddress);
    throw new UnauthorizedError('Invalid credentials');
  }

  // Check user is enabled
  if (!user.enabled) {
    throw new UnauthorizedError('Account is disabled. Contact system administrator.');
  }

  // Successful Login: Reset counters
  await prisma.user.update({
    where: { id: user.id },
    data: {
      failedLoginAttempts: 0,
      accountLocked: false,
      lockTime: null,
    },
  });

  const accessToken = generateAccessToken(user);
  const refreshToken = await createRefreshToken(user.id);

  logAudit(user.id, 'LOGIN_SUCCESS', 'Logged in successfully', ipAddress);

  return {
    id: user.id,
    username: user.username,
    email: user.email,
    roles: user.roles,
    accessToken,
    refreshToken,
    tokenType: 'Bearer',
    user: toUserDto(user),
  };
};

/**
 * Register Service
 */
export const registerUser = async (
  data: {
    username: string;
    email: string;
    passwordHash: string;
    firstName: string;
    lastName: string;
    phone?: string;
    roles?: Role[];
  },
  ipAddress?: string
) => {
  const existingUsername = await prisma.user.findUnique({
    where: { username: data.username },
  });
  if (existingUsername) {
    throw new BadRequestError('Username is already taken');
  }

  const existingEmail = await prisma.user.findUnique({
    where: { email: data.email },
  });
  if (existingEmail) {
    throw new BadRequestError('Email is already registered');
  }

  const hashedPassword = await bcrypt.hash(data.passwordHash, 10);
  const userRoles = data.roles && data.roles.length > 0 ? data.roles : [Role.PATIENT];

  const newUser = await prisma.user.create({
    data: {
      username: data.username,
      email: data.email,
      passwordHash: hashedPassword,
      firstName: data.firstName,
      lastName: data.lastName,
      phone: data.phone || null,
      roles: userRoles,
      enabled: true,
    },
  });

  // If PATIENT role is assigned, create Patient profile
  if (userRoles.includes(Role.PATIENT)) {
    await prisma.patient.create({
      data: {
        userId: newUser.id,
        firstName: newUser.firstName,
        lastName: newUser.lastName,
        email: newUser.email,
        phone: newUser.phone,
      },
    });
  }

  // If DOCTOR role is assigned, create Doctor profile
  if (userRoles.includes(Role.DOCTOR)) {
    await prisma.doctor.create({
      data: {
        userId: newUser.id,
        firstName: newUser.firstName,
        lastName: newUser.lastName,
        email: newUser.email,
        phone: newUser.phone,
        specialization: 'General',
        licenseNumber: `LIC-${newUser.id}-${Date.now().toString().slice(-4)}`,
        available: true,
      },
    });
  }

  logAudit(newUser.id, 'REGISTER_SUCCESS', `New user registered with roles: ${userRoles.join(', ')}`, ipAddress);

  return toUserDto(newUser);
};

/**
 * Token Refresh Service
 */
export const refreshUserToken = async (token: string, ipAddress?: string) => {
  const savedToken = await prisma.refreshToken.findUnique({
    where: { token },
    include: { user: true },
  });

  if (!savedToken) {
    throw new UnauthorizedError('Invalid refresh token');
  }

  if (new Date() > savedToken.expiryDate) {
    await prisma.refreshToken.delete({ where: { token } });
    throw new UnauthorizedError('Expired refresh token');
  }

  // Create new tokens (Token rotation)
  await prisma.refreshToken.delete({ where: { token } });

  const newAccessToken = generateAccessToken(savedToken.user);
  const newRefreshToken = await createRefreshToken(savedToken.user.id);

  return {
    accessToken: newAccessToken,
    refreshToken: newRefreshToken,
    tokenType: 'Bearer',
  };
};

/**
 * Logout Service
 */
export const logoutUser = async (userId: number, ipAddress?: string) => {
  const user = await prisma.user.findUnique({ where: { id: userId } });
  if (!user) {
    throw new NotFoundError('User not found');
  }

  await prisma.refreshToken.deleteMany({
    where: { userId },
  });

  logAudit(userId, 'LOGOUT_SUCCESS', 'Logged out successfully', ipAddress);
};

/**
 * Change Password Service
 */
export const changeUserPassword = async (
  userId: number,
  currentPass: string,
  newPass: string,
  ipAddress?: string
) => {
  const user = await prisma.user.findUnique({ where: { id: userId } });
  if (!user) {
    throw new NotFoundError('User not found');
  }

  const isMatch = await bcrypt.compare(currentPass, user.passwordHash);
  if (!isMatch) {
    throw new BadRequestError('Current password is incorrect');
  }

  const hashedNewPassword = await bcrypt.hash(newPass, 10);

  await prisma.$transaction([
    prisma.user.update({
      where: { id: userId },
      data: { passwordHash: hashedNewPassword },
    }),
    prisma.refreshToken.deleteMany({
      where: { userId },
    }),
  ]);

  logAudit(userId, 'PASSWORD_CHANGE', 'Password changed successfully', ipAddress);
};

/**
 * Get User by ID
 */
export const getUserById = async (id: number) => {
  const user = await prisma.user.findUnique({ where: { id } });
  if (!user) {
    throw new NotFoundError('User not found');
  }
  return toUserDto(user);
};

/**
 * Update User Roles
 */
export const updateRoles = async (userId: number, roles: Role[], ipAddress?: string) => {
  const user = await prisma.user.findUnique({ where: { id: userId } });
  if (!user) {
    throw new NotFoundError('User not found');
  }

  const updatedUser = await prisma.user.update({
    where: { id: userId },
    data: { roles },
  });

  logAudit(
    userId,
    'UPDATE_ROLES',
    `Roles updated to: ${roles.join(', ')}`,
    ipAddress
  );

  return toUserDto(updatedUser);
};

/**
 * Enable/Disable Account
 */
export const toggleStatus = async (userId: number, enabled: boolean, ipAddress?: string) => {
  const user = await prisma.user.findUnique({ where: { id: userId } });
  if (!user) {
    throw new NotFoundError('User not found');
  }

  const updatedUser = await prisma.user.update({
    where: { id: userId },
    data: { enabled },
  });

  const action = enabled ? 'ENABLE_USER' : 'DISABLE_USER';
  logAudit(userId, action, `User account ${enabled ? 'enabled' : 'disabled'}`, ipAddress);

  return toUserDto(updatedUser);
};

File: backend/auth-service/src/main/java/com/hospital/auth/service/AuthService.java
package com.hospital.auth.service;

import com.hospital.auth.dto.*;
import com.hospital.auth.entity.RefreshToken;
import com.hospital.auth.entity.Role;
import com.hospital.auth.entity.User;
import com.hospital.auth.repository.UserRepository;
import com.hospital.common.dto.PagedResponse;
import com.hospital.common.event.AuditEvent;
import com.hospital.common.exception.BusinessException;
import com.hospital.common.exception.ResourceNotFoundException;
import com.hospital.common.exception.UnauthorizedException;
import com.hospital.common.security.JwtTokenProvider;
import com.hospital.common.security.SecurityConstants;
import com.hospital.common.security.UserPrincipal;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Core authentication and user management service.
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider jwtTokenProvider;
    private final RefreshTokenService refreshTokenService;
    private final AuthenticationManager authenticationManager;
    private final RabbitTemplate rabbitTemplate;

    /**
     * Authenticate user and return JWT tokens.
     */
    @Transactional
    public AuthResponse login(LoginRequest request) {
        User user = userRepository.findByUsername(request.getUsernameOrEmail())
                .or(() -> userRepository.findByEmail(request.getUsernameOrEmail()))
                .orElseThrow(() -> new BadCredentialsException("Invalid credentials"));

        // Check account lock
        if (user.isAccountLocked()) {
            if (user.isLockExpired()) {
                user.resetFailedAttempts();
                userRepository.save(user);
            } else {
                throw new UnauthorizedException(
                        "Account is locked due to multiple failed login attempts. Try again after 30 minutes.");
            }
        }

        try {
            Authentication authentication = authenticationManager.authenticate(
                    new UsernamePasswordAuthenticationToken(
                            user.getUsername(),
                            request.getPassword()
                    )
            );

            // Reset failed attempts on successful login
            if (user.getFailedLoginAttempts() > 0) {
                user.resetFailedAttempts();
                userRepository.save(user);
            }

            UserPrincipal principal = (UserPrincipal) authentication.getPrincipal();
            String accessToken = jwtTokenProvider.generateAccessToken(principal);
            RefreshToken refreshToken = refreshTokenService.createRefreshToken(user);

            // Publish audit event
            publishAuditEvent(user, "LOGIN", "User logged in successfully");

            log.info("User '{}' logged in successfully", user.getUsername());

            return AuthResponse.of(accessToken, refreshToken.getToken(), toUserDto(user));

        } catch (BadCredentialsException e) {
            user.incrementFailedAttempts();
            userRepository.save(user);
            log.warn("Failed login attempt for user '{}'. Attempts: {}",
                    user.getUsername(), user.getFailedLoginAttempts());
            throw e;
        }
    }

    /**
     * Register a new user.
     */
    @Transactional
    public UserDto register(RegisterRequest request) {
        // Check for existing username/email
        if (userRepository.existsByUsername(request.getUsername())) {
            throw new BusinessException("USERNAME_EXISTS", "Username is already taken");
        }
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new BusinessException("EMAIL_EXISTS", "Email is already registered");
        }

        // Default to PATIENT role if none specified
        Set<Role> roles = request.getRoles();
        if (roles == null || roles.isEmpty()) {
            roles = Set.of(Role.PATIENT);
        }

        User user = User.builder()
                .username(request.getUsername())
                .email(request.getEmail())
                .passwordHash(passwordEncoder.encode(request.getPassword()))
                .firstName(request.getFirstName())
                .lastName(request.getLastName())
                .phone(request.getPhone())
                .enabled(true)
                .roles(new java.util.HashSet<>(roles))
                .build();

        User savedUser = userRepository.save(user);
        log.info("New user registered: {} with roles: {}", savedUser.getUsername(), roles);

        publishAuditEvent(savedUser, "REGISTER", "New user registered");

        return toUserDto(savedUser);
    }

    /**
     * Refresh the access token using a valid refresh token.
     */
    @Transactional
    public AuthResponse refreshToken(TokenRefreshRequest request) {
        RefreshToken refreshToken = refreshTokenService.validateRefreshToken(request.getRefreshToken());
        User user = refreshToken.getUser();

        UserPrincipal principal = buildUserPrincipal(user);
        String newAccessToken = jwtTokenProvider.generateAccessToken(principal);

        // Token rotation: create new refresh token
        RefreshToken newRefreshToken = refreshTokenService.createRefreshToken(user);

        return AuthResponse.builder()
                .accessToken(newAccessToken)
                .refreshToken(newRefreshToken.getToken())
                .tokenType("Bearer")
                .userId(user.getId())
                .username(user.getUsername())
                .build();
    }

    /**
     * Logout: revoke all refresh tokens for the user.
     */
    @Transactional
    public void logout(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", userId));
        refreshTokenService.revokeAllUserTokens(user);
        publishAuditEvent(user, "LOGOUT", "User logged out");
        log.info("User '{}' logged out", user.getUsername());
    }

    /**
     * Change password.
     */
    @Transactional
    public void changePassword(Long userId, ChangePasswordRequest request) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", userId));

        if (!passwordEncoder.matches(request.getCurrentPassword(), user.getPasswordHash())) {
            throw new BusinessException("INVALID_PASSWORD", "Current password is incorrect");
        }

        user.setPasswordHash(passwordEncoder.encode(request.getNewPassword()));
        userRepository.save(user);

        // Revoke all refresh tokens on password change
        refreshTokenService.revokeAllUserTokens(user);

        publishAuditEvent(user, "CHANGE_PASSWORD", "Password changed");
        log.info("Password changed for user: {}", user.getUsername());
    }

    /**
     * Get user by ID.
     */
    @Transactional(readOnly = true)
    public UserDto getUserById(Long id) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", id));
        return toUserDto(user);
    }

    /**
     * Get all users with pagination and optional search.
     */
    @Transactional(readOnly = true)
    public PagedResponse<UserDto> getAllUsers(int page, int size, String search, String sortBy) {
        Pageable pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, sortBy));
        Page<User> userPage;

        if (search != null && !search.isBlank()) {
            userPage = userRepository.searchUsers(search, pageable);
        } else {
            userPage = userRepository.findAll(pageable);
        }

        List<UserDto> users = userPage.getContent().stream()
                .map(this::toUserDto)
                .toList();

        return PagedResponse.of(
                users,
                userPage.getNumber(),
                userPage.getSize(),
                userPage.getTotalElements(),
                userPage.getTotalPages(),
                userPage.isFirst(),
                userPage.isLast()
        );
    }

    /**
     * Update user roles (admin only).
     */
    @Transactional
    public UserDto updateUserRoles(Long userId, Set<Role> roles) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", userId));

        user.setRoles(new java.util.HashSet<>(roles));
        User savedUser = userRepository.save(user);

        publishAuditEvent(savedUser, "UPDATE_ROLES",
                "Roles updated to: " + roles.stream().map(Enum::name).collect(Collectors.joining(",")));

        return toUserDto(savedUser);
    }

    /**
     * Enable/disable a user account.
     */
    @Transactional
    public UserDto toggleUserStatus(Long userId, boolean enabled) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", userId));

        user.setEnabled(enabled);
        User savedUser = userRepository.save(user);

        String action = enabled ? "ENABLE_USER" : "DISABLE_USER";
        publishAuditEvent(savedUser, action, "User account " + (enabled ? "enabled" : "disabled"));

        return toUserDto(savedUser);
    }

    // ==================== Helpers ====================

    private UserDto toUserDto(User user) {
        return UserDto.builder()
                .id(user.getId())
                .username(user.getUsername())
                .email(user.getEmail())
                .firstName(user.getFirstName())
                .lastName(user.getLastName())
                .phone(user.getPhone())
                .enabled(user.isEnabled())
                .roles(user.getRoles().stream().map(Enum::name).collect(Collectors.toSet()))
                .createdAt(user.getCreatedAt())
                .updatedAt(user.getUpdatedAt())
                .build();
    }

    private UserPrincipal buildUserPrincipal(User user) {
        return UserPrincipal.builder()
                .id(user.getId())
                .username(user.getUsername())
                .email(user.getEmail())
                .password(user.getPasswordHash())
                .firstName(user.getFirstName())
                .lastName(user.getLastName())
                .enabled(user.isEnabled())
                .roles(user.getRoles().stream()
                        .map(role -> "ROLE_" + role.name())
                        .collect(Collectors.toList()))
                .build();
    }

    private void publishAuditEvent(User user, String action, String description) {
        try {
            AuditEvent event = AuditEvent.builder()
                    .userId(user.getId())
                    .username(user.getUsername())
                    .action(action)
                    .entityType("USER")
                    .entityId(user.getId())
                    .serviceName("auth-service")
                    .description(description)
                    .severity(AuditEvent.AuditSeverity.INFO)
                    .timestamp(LocalDateTime.now())
                    .build();
            rabbitTemplate.convertAndSend(
                    SecurityConstants.EXCHANGE_HOSPITAL,
                    SecurityConstants.ROUTING_KEY_AUDIT,
                    event
            );
        } catch (Exception ex) {
            log.error("Failed to publish audit event: {}", ex.getMessage());
        }
    }
}
